frontend/src/app/shared/pipes/proof.pipe.ts

```diff
@@ -0,0 +1,56 @@
+import { Pipe, PipeTransform } from "@angular/core";
+import { DomSanitizer, SafeHtml } from "@angular/platform-browser";
+
+const SYMBOL_MAPPING = {
+    "▵": "△",
+    "▾": "▼",
+    "▴": "▲",
+    "▿": "▽",
+    "◇": "♢",
+    "□": "□",
+};
+
+@Pipe({
+    name: "proof",
+})
+export class ProofPipe implements PipeTransform {
+    transform(value: string): SafeHtml {
+
+        // Object.keys() is not properly typed.
+        const originalSymbols = Object.keys(
+            SYMBOL_MAPPING,
+        ) as (keyof typeof SYMBOL_MAPPING)[];
+
+        for (const symbol of originalSymbols) {
+            // Wrap all text in between a symbol and '(' with a <sup> tag.
+            value = value.replace(
+                new RegExp(`(${symbol})(.*?)\\(`, "g"),
+                (match, p1, p2) => {
+                    return `${p1}<sup>${p2}</sup>(`;
+                },
+            );
+
+            // Replace symbols according to the SYMBOL_MAPPING object.
+            value = value.replace(
+                new RegExp(symbol, "g"),
+                SYMBOL_MAPPING[symbol],
+            );
+        }
+
+        // Replace arrows with implicature symbols
+        value = value.replace(/⟶/g, " ⊸ ");
+
+        // Wrap all variables (i.e. instances of 'c' followed by one or multiple numbers) in a <var> tag.
+        value = value.replace(/(c\d+)/g, '<var class="variable">$1</var>');
```

Comment:
since this is something we create in angular and doesn't come from the server, I think it makes more sense to format correctly directly in the template, instead of using a pipe